### DISCUSSION TITLE
Execute code action programatically

### DISCUSSION THREAD

**Discussion by Andre0991** [joaotavora/eglot#983] [Q&A] [answered]
Hello.

The clojure LSP server has a "Clean namespace" code action. I'd like to call use code action programmatically, because I'm creating some custom code for formatting the buffer when I switch windows:

```elisp
(defun format-if-clojure
    (&rest _)
  (when (and buffer-file-name
	     (buffer-modified-p)
	     (derived-mode-p 'clojure-mode)
	     (bound-and-true-p eglot-managed-mode-hook))
    (eglot-format)))

(advice-add 'other-window :before #'format-if-clojure)
```

As the snippet above shows, I'm already using `eglot-format`, but I'd like to also use the 'clean ns' code action, which complements it.

Is that possible?

Just for reference, `lsp-mode` provides commands for language-specific code actions. For example, it has `lsp-clojure-clean-ns` ([implementation](https://github.com/emacs-lsp/lsp-mode/blob/1d9ec3f30a6491b58a904b071694a1a8a0faca87/clients/lsp-clojure.el#L126)).

I suppose this is outside `eglot`'s scope, as it tries to be as language agnostic as possible, but I'm wondering how I could add something similar to that to my own config.

**Comment by joaotavora** [comment, marked as answer]
Code actions available from the server can be requested using `M-x eglot-code-actions`.  See its doc. A simple call with no region selected will interrogate the server and ask for a code action at the current point.  But if you select a region, it will ask the server for all available code actions in the region. See if that convinces the server to offer you the action you're interested in.  To make an `andre/clojure-clean-ns` command (naming suggestion based on your username)  would then entail calling `eglot-code-actions` with suitable `beg` and `end` arguments.  Maybe `point-min` and `point-max`, I suppose.

**Reply by Andre0991** [reply]
(I deleted my previous message for reducing noise in this discussion)

Oh, never mind. I just read the docstring and I see that it gets code actions as an argument. Sorry for the noise.

Now, I'm trying to find out what's the actual code action name – I tried "clean-ns" (just like `lsp-mode` code) and "Clean namespace" (as it's displayed in the list) and I'm getting this error:

```
(eglot-code-actions (point-min) (point-max) "Clean namespace")
```

```
Debugger entered--Lisp error: (error "[eglot] No \"Clean namespace\" code actions here")
  error("[eglot] %s" "No \"Clean namespace\" code actions here")
  eglot--error("No \"%s\" code actions here" "Clean namespace")
  apply(eglot--error ("No \"%s\" code actions here" "Clean namespace"))
  eglot-code-actions(1 6227 "Clean namespace")
```

How do I know the actual code action name?

**Reply by joaotavora** [reply]
Use `M-x eglot-events-buffer` and look around for the description of the desired action and its `:kind` key.  It's most likely a string. Then put that string as the `action-kind` argument of `eglot-code-actions`.

**Reply by Andre0991** [reply]
Ohhh. In the end, I've found that the kind is `"source.organizeImports"`, which is already covered by `eglot-code-action-organize-imports` :)

Thanks!

**Reply by joaotavora** [reply]
ha! good one!
